### ISSUE TITLE
@apply inside @define-mixin (postcss-mixins) throw "No parent could be found" error

### ISSUE THREAD

**Issue by hyanmandian** [tailwindlabs/tailwindcss#4114]
### What version of Tailwind CSS are you using?

2.1.1

### What build tool (or framework if it abstracts the build tool) are you using?

postcss 8.1.14 and rollup via TSDX

### What version of Node.js are you using?

15.8.0

### What browser are you using?

N/A

### What operating system are you using?

Linux (Ubuntu)

### Reproduction repository

https://github.com/hyanmandian/tailwind-postcss-bug

### Describe your issue

Basically, when I'm using Postcss with Tailwind and Postcss Mixins plugin together I receive "No parent could be found" error. In the reproduction repository, you can check what is happening.

**Comment by mammadataei** [comment]
The same issue with `postcss-nested`.

**Comment by hyanmandian** [comment]
I found a workaround to solve that by wrapping with an empty parent selector, something like:

```postcss
@define-mixin example {
  & {
    @apply text-white;
  }
}
```

**Comment by d4h0** [comment]
I'm not sure if this is the same issue, but I get the same error when I try to use `@apply` within a Paged Media query `@page {@bottom-right { /* here */ }}`.

I couldn't find a good workaround so far (the empty parent selector workaround didn't work for me).

What I do for now, is to use the [theme function](https://tailwindcss.com/docs/functions-and-directives#theme).

But using Tailwind classes in my Print CSS would be awesome!

**Comment by anantakrishna** [comment]
@hyanmandian are you sure this line is correct?

https://github.com/hyanmandian/tailwind-postcss-bug/blob/bbc9340fe9a046231425593ea58a8438fbf05b1d/src/index.css#L22

I believe it should be

```
.dont-work { @mixin not-working-mixin; }
```

**Comment by hyanmandian** [comment]
Thanks for letting me know, I just fixed this typo error.

**Comment by Nagell** [comment]
For me it was about postcss config and a 'bug' in tailwind
and something like this was a solution:

```js
let postcss = require('postcss')

module.exports = {
  plugins: [
    {
      postcssPlugin: 'grouped',
      Once(root, { result }) {
        return postcss([
          require('postcss-import'),
          require('postcss-mixins'),
          require('postcss-simple-vars'),
        ]).process(root, result.opts)
      },
    },
    require('tailwindcss'),
    require('postcss-nested'),
    require('autoprefixer'),
  ],
}
```

Grabbed from: https://giters.com/tailwindlabs/tailwindcss/issues/3931

I hope it will help. Good luck guys!

**Comment by Erin69** [comment]
I couldn't find a good workaround so far the empty parent selector workaround didn't work for me.

**Comment by adamwathan** [comment]
Hey! We've looked into this and the root cause is that in PostCSS, plugins that use the visitor API always run _after_ plugins that don't use the visitor API, rather than in the order you specify in your `postcss.config.js` file. This is tricky because as you have all noticed, `postcss-mixins` needs to run _before_ Tailwind, because Tailwind doesn't understand the syntax that `postcss-mixins` wants you to write until it has been transformed into normal CSS.

Unfortunately there's nothing we can do about this inside Tailwind to make this work, and Tailwind isn't a good candidate for migrating to the visitor API.

So going to close this one I'm afraid, and recommend the workaround here for anyone who has this issue:

https://github.com/tailwindlabs/tailwindcss/issues/4114#issuecomment-989931186

That will force `postcss-mixins` to run as if it were not a visitor plugin, which gives you the control you need to make sure it runs before Tailwind instead of after.
